File: Backtracking/93_RestoreIPAddresses.py
import copy


class Solution(object):

    def restoreIpAddresses(self, s):
        """
        :type s: str
        :rtype: List[str]
        """
        lst = []
        self.backtract(lst, [], s, 0)
        return lst

    def backtract(self, lst, tmplist, s, start):
        # print tmplist
        if start == len(s) and len(tmplist) == 4:
            lst.append(".".join(copy.deepcopy(tmplist)))
        elif start < len(s) and len(tmplist) < 4:
            for i in range(1, 4):
                if start + i > len(s):
                    continue
                tmp = s[start:start + i]
                if self.fit_ip_address(tmp):
                    # print tmp
                    tmplist.append(tmp)
                    self.backtract(lst, tmplist, s, start + i)
                    tmplist.pop()

    def fit_ip_address(self, s):
        if len(s) == 0 or len(s) >= 4:
            return False
        if len(s) > 1 and s.startswith('0'):
            return False
        if int(s) < 0 or int(s) > 255:
            return False
        return True

File: Backtracking/93_RestoreIPAddresses.java
public class Solution {
    public List<String> restoreIpAddresses(String s) {
        List<String> lst = new ArrayList<>();
        restoreIp(lst, new ArrayList<String>(), s, 0);
        return lst;
    }

    public void restoreIp(List<String> lst, List<String> tmp, String s, int start) {
        if (start == s.length() && tmp.size() == 4) {
            lst.add(String.join(",", tmp));
        } else if (start < s.length() && tmp.size() < 4) {
            for (int i=1; i<=4; i++) {
                String ele = s.substring(start, start+i);
                if ((ele.startwith('0') && ele.length()>1) || (ele.length()==3 && Integer.parseInt(ele)>255)) {
                    continue;
                }
                tmp.add(ele);
                restoreIp(lst, tmp, s, start+i);
                tmp.remove(tmp.size()-1);
            }
        }
    }
}